src/main.rs

```diff
@@ -139,204 +174,209 @@ impl DiemDemo {
+                        if let Err(_) = self.init_state(None).await {
+                            return Err(Error::FailedToInitState);
+                        }
+
+                        state_initiated = true;
+                    }
+
+                    for event in new_events {
+                        if let Ok(transaction) = self.get_transaction_by_version(event.transaction_version) {
+                            println!("received transaction:{:?}", transaction);
+                            if let Ok(transaction_with_proof) = self.get_transaction_proof(&transaction) {
+                                println!("transaction_with_proof:{:?}", transaction_with_proof);
+
+                                let transaction_with_proof_b64 = base64::encode(&bcs::to_bytes(&transaction_with_proof).unwrap());
+                                let _resp = pr.query(DIEM_CONTRACT_ID, QueryReqData::VerifyTransaction
+                                    { account_address: account_address.clone(), transaction_with_proof_b64 }).await?;
+                            } else {
+                                println!("get_transaction_proof error");
+                            }
+                        } else {
+                            println!("get_transaction_by_version error");
+                        }
                     }
+
+                    self.received_events = Option::from(received_events);
+                } else {
+                    println!("get receiving events error");
                 }
+            }
 
-                // Init transactions
-                let mut batch = JsonRpcBatch::new();
-                batch.add_get_account_transactions_request(self.account.as_ref().unwrap().address.clone(),0, self.account.as_ref().unwrap().sequence_number.clone(), true);
-                let responses = self.rpc_client.execute(batch).unwrap();
-                match get_response_from_batch(0, &responses).unwrap().as_ref() {
-                    Ok(resp) => {
-                        self.transactions = Option::from(TransactionView::vec_from_response(resp.clone()).unwrap());
-                    },
-                    Err(_) => {
-                        self.transactions = None;
+            // Init transactions
+            let mut batch = JsonRpcBatch::new();
+            batch.add_get_account_transactions_request(self.account.as_ref().unwrap().address.clone(),
+                0, self.account.as_ref().unwrap().sequence_number.clone(), false);
+            if let Ok(resp) = self.request_rpc(batch) {
+                let mut need_sync_transactions: Vec<TransactionView> = Vec::new();
+                let transactions = TransactionView::vec_from_response(resp.clone()).unwrap();
+                for transaction in transactions.clone() {
+                    let exist = self.transactions.as_ref().is_some()
+                        && self.transactions.as_ref().unwrap().iter().any(|x| x.version == transaction.version);
```

Comment:
Similarly, consider use `HashMap` or `HashSet` to remember which tx is already synced.